feat: Add support to specify file permissions for pvc hostpaths

sushiMix opened the pull request:
## Pull Request template

**Why is this PR required? What issue does it fix?**:
Fixes https://github.com/openebs/dynamic-localpv-provisioner/issues/95.

**What this PR does?**:
It allows to configure default directory mode using LocaMode entry in cas.openebs.io/config.

**Does this PR require any upgrade changes?**: No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    cas.openebs.io/config: |
      - name: StorageType
        value: "hostpath"
      - name: BasePath
        value: "/opt/data/"
      - name: HostPathMode
        value: "770"
    openebs.io/cas-type: local
  name: local-path
provisioner: openebs.io/local
reclaimPolicy: Delete
volumeBindingMode: WaitForFirstConsumer
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: task-pv-claim
spec:
  storageClassName: local-path
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: task-pv-pod
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
        claimName: task-pv-claim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
        - mountPath: "/usr/share/nginx/html"
          name: task-pv-storage
```

check the file mode of the created folder (0770 instead of 0777)

**Any additional information for your reviewer?** : 
_Mention if this PR is part of any design or a continuation of previous PRs_


**Checklist:**
- [x] Fixes #95
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated? 
- [] Commit has unit tests
- [] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track: 
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them: 

**PLEASE REMOVE BELOW INFORMATION BEFORE SUBMITTING**

IMPORTANT: Please review the [CONTRIBUTING.md](../CONTRIBUTING.md) file for detailed contributing guidelines.

sushiMix committed:
feat: Add support to specify file permissions for pvc hostpaths

Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix commented:
Sorry for last PR with a bad rebase when signing.

niladrih reviewed (commented):
Hi @sushiMix. There is already one file-mode/file-mode API in place over at the [openebs/dynamic-nfs-provisioner repo](https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/docs/tutorial/file-permissions.md).
PR link: https://github.com/openebs/dynamic-nfs-provisioner/pull/125

While I don't have any issues with this implementation, it'd be useful to have a stable cas-config API across openebs storage engines. Also, the above implemetation supports file-permissions as well, which may be useful to your use-case.

sushiMix commented:
Hello @niladrih I did a test with storage class and then pvc annotation configuration.

After I did rebase and squashed the commits.

I only support the mode changed. The user and group are left to kubelet behavior (FsGroup is supported for local volume in kubernetes). Need a bit more code change to update the command sent.

niladrih commented:
Great! Could you also add a documentation .md file in the docs directory with instructions on using this feature?

sushiMix commented:
> Great! Could you also add a documentation .md file in the docs directory with instructions on using this feature?

I updated the QuickStart, do you need a more detailed documentation?

@niladrih I added a specific page.

niladrih commented on cmd/provisioner-localpv/app/config.go:
```diff
@@ -170,15 +190,21 @@ func (p *Provisioner) GetVolumeConfig(ctx context.Context, pvName string, pvc *c
+	if len(strings.TrimSpace(pvcCASConfigStr)) != 0 {
+		pvcCASConfig, err := cast.UnMarshallToConfig(pvcCASConfigStr)
+		if err == nil {
+			pvConfig = cast.MergeConfig(pvcCASConfig, pvConfig)
```
Here pvcCASConfig is essentially the scCASConfig from the previous block. The SC config (cluster admin's configuration) should have precedence over the PVC config (cluster user's configuration). User shouldn't be able to override admin's config.

```suggestion
			pvConfig = cast.MergeConfig(pvConfig, pvcCASConfig)
```

niladrih commented on cmd/provisioner-localpv/app/config.go:
```diff
@@ -130,6 +130,26 @@ const (
+	// This is the cas-template key for all file permission 'data' keys
+	KeyFilePermissions = "FilePermissions"
+
+	// FsUID defines the user owner of the shared directory
+	KeyFsUID = "UID"
+
+	// FsGID defines the group owner of the shared directory
+	KeyFsGID = "GID"
```
These constants could be omitted as they are not used.

niladrih commented on cmd/provisioner-localpv/app/config.go:
```diff
@@ -338,18 +368,83 @@ func (c *VolumeConfig) IsExt4QuotaEnabled() bool {
+	return permissionEnabledQuotaBool
+}
+
+// GetFsGID fetches the group owner's ID from
+// PVC annotation, if specified
+// NOT YET USED
+func (c *VolumeConfig) GetFsGID() string {
+	if c.IsPermissionEnabled() {
+		configData := c.getData(KeyFilePermissions)
+		if configData != nil {
+			if val, p := configData[KeyFsGID]; p {
+				return strings.TrimSpace(val)
+			}
+		}
+	}
+	return ""
+}
+
+// GetFsGID fetches the user owner's ID from
+// PVC annotation, if specified
+// NOT YET USED
+func (c *VolumeConfig) GetFsUID() string {
+	if c.IsPermissionEnabled() {
+		configData := c.getData(KeyFilePermissions)
+		if configData != nil {
+			if val, p := configData[KeyFsUID]; p {
+				return strings.TrimSpace(val)
+			}
+		}
+	}
+	return ""
+}
```
These need not be a part of this PR, as these are not used.

niladrih commented on cmd/provisioner-localpv/app/config.go:
```diff
@@ -338,18 +368,83 @@ func (c *VolumeConfig) IsExt4QuotaEnabled() bool {
+		}
+	}
+	//Keep the original default mode
+	return "0777"
```
The default value should not be set here. This function should default to an empty value if no valid configuration is found. The default value should be set when this function is used.
```suggestion
	return ""
```

niladrih commented on cmd/provisioner-localpv/app/provisioner_hostpath.go:
```diff
@@ -76,7 +77,7 @@ func (p *Provisioner) ProvisionHostPath(ctx context.Context, opts pvController.P
 
 	//Before using the path for local PV, make sure it is created.
-	initCmdsForPath := []string{"mkdir", "-m", "0777", "-p"}
+	initCmdsForPath := []string{"mkdir", "-m", volumeConfig.GetFsMode(), "-p"}
```
```suggestion
        fsMode := volumeConfig.GetFsMode()
        // Set default value if FilePermissions mode is not specified.
        if len(fsMode) == 0 {
                fsMode = "0777"
        }
	initCmdsForPath := []string{"mkdir", "-m", fsMode, "-p"}
```

niladrih commented on docs/tutorials/hostpath/filepermissions.md:
```diff
@@ -0,0 +1,54 @@
+Hostpath LocalPV will by default create folder with the following rights: `0777`. In some usecases, these rights are too wide and should be reduced.
+As an important point, when using hostpath the underlying PV will be a localpath whichs allows kubelet to chown the folder based on the [fsGroup](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods))
+
+We allow to reduce this filepermission using :
```
nit:
```suggestion
We allow to set file permissions using:  
```

niladrih commented on docs/tutorials/hostpath/filepermissions.md:
```diff
@@ -0,0 +1,54 @@
+
+With such configuration the folder will be crated with `0770` rights for all the PVC using this storage class.
+
+The same configuration is available at PVC level to have a more fined grained configuration capability (overrding the Storage class configuration level):
```
We shouldn't override SC config, as it may be an admin's directive. The PVC config should not be preferred when the admin has specified a config.

sushiMix replied:
done

niladrih reviewed (commented):
Thank you for the changes @sushiMix! I have left some comments.

niladrih reviewed (commented):
Additionally, the enabled key is not required here. Mentioning the details in 'data' is good enough.

sushiMix committed:
Update cmd/provisioner-localpv/app/config.go

Co-authored-by: Niladri Halder <niladri.halder26@gmail.com>
Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix committed:
Update docs/tutorials/hostpath/filepermissions.md

Co-authored-by: Niladri Halder <niladri.halder26@gmail.com>
Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix committed:
Update cmd/provisioner-localpv/app/provisioner_hostpath.go

Co-authored-by: Niladri Halder <niladri.halder26@gmail.com>
Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix committed:
Update cmd/provisioner-localpv/app/config.go

Co-authored-by: Niladri Halder <niladri.halder26@gmail.com>
Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix committed:
review

Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix committed:
remove enabled flag

Signed-off-by: sushiMix <53741704+sushiMix@users.noreply.github.com>

sushiMix commented:
@niladrih Happy New Year, I had time to perform code modification.

niladrih commented:
Hi @sushiMix. Could you add ginkgo tests to this PR to test your feature? You could use https://github.com/openebs-archive/dynamic-nfs-provisioner/blob/develop/tests/nfs_server_file_permissions_test.go as an example in this case.

Let me know if you need help with this.

Also, I am sorry I haven't been able to review this!

niladrih commented:
@sushiMix I've pulled in your commits to #264. Also I've fixed a bug where the mode is set to unexpected values if there are no "" surrounding the value on the storageClass/pvc.

niladrih commented:
Closing this PR.
